### ISSUE TITLE
Early termination after CEGUI::UnknownObjectException

### ISSUE THREAD

**Issue by t4im** [OpenDungeons/OpenDungeons#965]
OpenDungeons terminates for me when pressing any button in the mainmenu (with the exception of the button used to open the Settings dialog) with a CEGUI version newer than 0.8.4

Tested with ee92d4440
Debian Jessie amd64
With a recent build of cegui v0-8 (5967:f68564f1404e)
(tag v0-8-4 seems to work fine)

```
CEGUI::UnknownObjectException in function 'CEGUI::NamedElement* CEGUI::NamedElement::getChildElement(const CEGUI::String&) const' (../src/cegui/cegui/src/NamedElement.cpp:150) : The Element object referenced by '__auto_titlebar__' is not attached to Element at 'SettingsWindow'.
terminate called after throwing an instance of 'CEGUI::UnknownObjectException'
  what():  CEGUI::UnknownObjectException in function 'CEGUI::NamedElement* CEGUI::NamedElement::getChildElement(const CEGUI::String&) const' (../src/cegui/cegui/src/NamedElement.cpp:150) : The Element object referenced by '__auto_titlebar__' is not attached to Element at 'SettingsWindow'.
Aborted
```

Full [CEGUI.log](https://gist.github.com/t4im/834e72d0112c6351a0e0)

**Comment by Bertram25** [comment]
@t4im Mind if I add a topic on CEGUI forums? ident and the others will easily spot the bug IMHO.

**Comment by akien-mga** [comment]
Note that @t4im edited the first topic to say that 0.8.4 works fine, so we'll have to see with the CEGUI guys if it's an upstream regression in the v0-8 branch (upcoming 0.8.5), or if it's a wanted change and we'll have to adapt to it.

**Comment by Bertram25** [comment]
> so we'll have to see with the CEGUI guys if it's an upstream regression in the v0-8 branch (upcoming 0.8.5), or if it's a wanted change and we'll have to adapt to it.

Completely agreed, it's done: http://cegui.org.uk/forum/viewtopic.php?f=10&t=7111

I'll keep an eye on that one.

**Comment by Bertram25** [comment]
Setting this to later, until this is sorted out with the Cegui team.

**Comment by Bertram25** [comment]
The Cegui team answered and spoke about a children initialization order problem on their own. Don't know much more, though. I'll ask if they need help.

**Comment by JKatzwinkel** [comment]
I can build with cegui `0.8.4-12` and run the game without problems. The `UnknownObjectException` persists if I build using cegui `0.8.5-1`.

**Edit:** (_meaning arch linux package releases_)

**Comment by akien-mga** [comment]
I get the package release numbers you mentioned come from Ubuntu? Keep in mind that there are tons of Linux distros so 0.8.4-12 does not mean much in itself.

I've noticed issues with the newly released 0.8.5 too though, I'll have to investigate. I like how the CEGUI team manages to break compatibility at each bugfix release...

**Comment by akien-mga** [comment]
@OpenDungeons/developers If there's a simple fix to make it work with both CEGUI <= 0.8.4 and >= 0.8.5, it would be nice to get it before releasing 0.7.0, as I can already see coming the PlayDeb guy updating CEGUI without testing OD against it, and us getting bug reports about our game being broken :p

Additionally, I've pushed CEGUI 0.8.5 to Mageia's repos already too, so my official opendungeons package is probably broken too. If there's no solution compatible with both CEGUI versions, I can just add the needed patch in my Mageia package though.

**Comment by hwoarangmy** [comment]
I confirm I also have the issue with CEGUI 0.8.5. I guess it comes from the way we set the window title (according to the exception). There might be a better way than to get the child window called "**auto_titlebar**". Note that we might also try to get rid of "**auto_closebutton**" which is much more used and might have the same problem. My guess is that "__auto_*" are not meant to be directly adressed from code. There might be some other event triggered when the windows is closed. I might have a look to check.

**Comment by hwoarangmy** [comment]
I've had a look and it seems not too hard to fix. It seems the proper way to set a windows title is calling window->setText.
Concerning **auto_closebutton**, it seems the proper way is to subscribe to "CEGUI::FrameWindow::EventCloseClicked". I will have a look at the remaining "auto*".
Once it is done, I will release a PR and you can try if it works with CEGUI 0.8.5.

**Comment by Bertram25** [comment]
> Concerning auto_closebutton, it seems the proper way is to subscribe to "CEGUI::FrameWindow::EventCloseClicked". I will have a look at the remaining "auto*".
> Once it is done, I will release a PR and you can try if it works with CEGUI 0.8.5.

Interesting :D Where did you get this information? Does this mean we could get rid of auto stuff in the code or in the looknfeels?

**Comment by hwoarangmy** [comment]
For setting the window title:
http://cegui.org.uk/wiki/The_Beginner_Guide_to_Creating_a_CEGUI_Window

For closing the window, I asked google if there was an event and it answered me:
http://cegui.org.uk/forum/viewtopic.php?t=1901

Note that I have had a look in the look n feel and it seems the auto\* are defined in our own widget declarations. Because of that, I'm not really sure the bug will be fixed. However, I have already tested replacing auto_closebutton with FrameWindow::EventCloseClicked and auto_titlebar with setText and it works. In any case, I think it is better to rely on such events rather than adding listeners to child events depending on the look and field.

**Comment by Bertram25** [comment]
> In any case, I think it is better to rely on such events rather than adding listeners to child events depending on the look and field.

Agreed. auto window have revealed being a muddy thing to deal with. Plus, they get messed up when editing windows with CEED. So if we have to remove them, all the better.

**Comment by hwoarangmy** [comment]
I've had a closer look and seen some stuff. First, the __auto\* are alo hardcoded in CEGUI code and defined in OD look n feel (and it is required by CEGUI). Thus, calling them as it is done doesn't seem to be bad after all. However, there are higher level events (at Frame level) and I think it is better to use those instead of dealing directly with the child widgets so I will create a PR with corresponding changes.

Concerning the bug itself, IMHO, it is a bug in CEGUI. I have updated the issue in their forum with what I've seen and we will see how it turns. So, to answer @akien-mga, unless the CEGUI team can think about a workaround, I don't think we can keep compatibility with 0.8.5.

**Comment by hwoarangmy** [comment]
Note that it seems the CEGUI team has fixed the bug. I will try ASAP the fix. Moreover, they seem to want to go for another release if the bug is fixed.

**Comment by hwoarangmy** [comment]
I've just tried the CEGUI fix and I don't get the crash anymore. Let's wait if they go for another release or not and we will see ^^

**Comment by Bertram25** [comment]
Perfect! :)

**Comment by JKatzwinkel** [comment]
Very nice!

**Comment by akien-mga** [comment]
BTW, with the recent changes to window titles, etc., are we still compatible with 0.8.4 or do we need 0.8.5 + the fix? And subsequent question, what version should I use for the Linux binaries for 0.7.0?

**Comment by hwoarangmy** [comment]
> BTW, with the recent changes to window titles, etc., are we still compatible with 0.8.4 or do we need 0.8.5 + the fix? And subsequent question, what version should I use for the Linux binaries for 0.7.0?

We are compatible with 0.8.4 and 0.8.5 + fix. The problem comes from the CEGUI FrameWindow core code and makes our code not compatible with official 0.8.5.
So, to answer the second question, you should use 0.8.4 until the 0.8.6 is out ^^

**Comment by hwoarangmy** [comment]
@akien-mga Note that CEGUI 0.8.6 is out. I've not tried it yet but I guess it fixes our bug. I will try it ASAP but probably not before tomorrow night. If you test on Linux, please let us know if it works well and if you see the wandering CEGUI artefact or not anymore.

**Comment by akien-mga** [comment]
Thanks for the notice, I'll update my Mageia package as well as the version I use for OD binaries. Based on the changelog, I don't expect many good surprises though, it should be just like 0.8.5 but without our issue. But maybe the 0.8.5 version brought useful fixes, we'll see :)

**Comment by akien-mga** [comment]
Oh my... They changed their soname again for a bugfix release that is just supposed to fix a couple minor bugs. I think I never saw a lib with that bad a soname history as CEGUI...

The libs used to be `libCEGUI*-0.so.2`, now it's `libCEGUI*-0.so.1.5.1` (what of course any computer would consider as a _lower_ version, but whatever...). I'll still make a bug report upstream to make sure they know what they're doing, they appear to always change the library major at random, and especially _not_ change it when they actually do break the ABI.

**Comment by akien-mga** [comment]
Done: https://bitbucket.org/cegui/cegui/issues/1123/library-soname-was-changed-from-2-to-1-in

Doesn't impact much for us apart from the fact that binaries build for CEGUI 0.8.5 won't find the library they're looking for if you have 0.8.6, but as a packager I had to rant :p

**Comment by akien-mga** [comment]
BTW I guess we can close this issue with resolution: _don't use CEGUI 0.8.5_?

**Comment by hwoarangmy** [comment]
agreed. The way they name their releases is their choice...

**Comment by akien-mga** [comment]
Just for the reference, upstream released CEGUI 0.8.7 with a fix for the bogus soname: http://cegui.org.uk/download/cegui-087

**Comment by Bertram25** [comment]
Excellent! :)
